### PR TITLE
Implemented Hashable trait for ProtocolState

### LINKED ISSUE #166: [M2] Implement Poseidon Hash for State Hash
**Issue summary**
Implement Poseidon Hash for State Hash, Currently it uses blake2b Hash, also add test case to validate tiebreaker logic in select longer chain method

**Other information and links**
https://github.com/ChainSafe/mina-rs/blob/4d2a01611f1bed2638289a1c5751d56239bf9769/consensus/src/common.rs#L100-L104

API and framework for Mina hashing. It is a safe wrapper around Mina’s instances of the [Poseidon arithmetic sponge](https://github.com/o1-labs/cryptography-rfcs/blob/master/mina/001-poseidon-sponge.md) that converts it from a sponge into a hash interface.

### PR DESCRIPTION
**Pull request by sudo-shashank**
**Summary of changes**
Changes introduced in this pull request:
- [Implemented hashable trait for ProtocolState and Used Kimchi Hasher](https://github.com/ChainSafe/mina-rs/pull/211/commits/0d6b5aa7e963c994e0601dad4540a06ac6fb226a)
- [Added select longer chain tie breaker logic test which use state hash comparison](https://github.com/ChainSafe/mina-rs/pull/211/commits/0d6b5aa7e963c994e0601dad4540a06ac6fb226a)

**Reference issue to close (if applicable)**

Closes #166 

**Other information and links**

### CODE REVIEW

File: consensus/src/common.rs [hunk 1]
```diff
@@ -120,21 +121,16 @@ impl Chain<ProtocolState> for ProtocolStateChain {
...
+            hasher.hash(&s.body.consensus_state.last_vrf_output)
         })
     }
 
-    // FIXME: this currently returns a blake2b hash. It should
-    // return a poseidon hash according to: https://github.com/MinaProtocol/mina/blob/32f529d8d8d712a44ee75be66061ce08cbdc8924/docs/specs/consensus/README.md#517-hashstate
-    fn state_hash(&self) -> Option<StateHash> {
-        self.top().map(|s| s.hash())
+    fn state_hash(&self) -> Option<Fp> {
+        let mut hasher = create_kimchi::<ProtocolState>(());
```
Comment (hanabi1224) [inline consensus/src/common.rs:132]: It's expensive to create a new hasher every time, better to create a pool and reuse the hasher.

**Review by hanabi1224 — APPROVED** [review]
Nice work! Approved with some suggestions

File: proof-systems-shim/src/lib.rs [hunk 2]
```diff
@@ -14,6 +14,7 @@
 //! ```
 //!
 
+pub use mina_curves;
```
Comment (hanabi1224) [inline proof-systems-shim/src/lib.rs:17]: You don't need to change this crate, Fp is re-exported from mina_hasher

File: proof-systems-shim/Cargo.toml [hunk 3]
```diff
@@ -7,6 +7,8 @@ edition = "2021"
 license = "Apache-2.0"
 
 [dependencies]
-mina-hasher = {git = "https://github.com/o1-labs/proof-systems", rev = "9816b804fe2e8336fbe20d2e3d977d766582c4f0"}
```
Comment (hanabi1224) [inline proof-systems-shim/Cargo.toml:10]: Please also revert the changes here, `Fp` is available under `mina-hasher`

File: proof-systems-shim/src/lib.rs (hunk 2 above)
Comment (sudo-shashank) [reply proof-systems-shim/src/lib.rs]: fixed

File: base/src/consensus_state.rs [hunk 4]
```diff
@@ -80,6 +93,36 @@ pub struct ConsensusState {
     pub supercharge_coinbase: bool,
 }
 
+impl Hashable for ConsensusState {
```
Comment (creativcoder) [inline base/src/consensus_state.rs:96]: This impl block is mostly same except for the function body of `to_roinput`. Can we write a readable macro to remove the repetitive parts? I believe you could just use a token tree `tt` for the function body.

File: base/src/consensus_state.rs [hunk 5]
```diff
@@ -18,6 +20,11 @@ use smart_default::SmartDefault;
 #[derive(Clone, Default, PartialEq, Debug, From, Serialize)]
 pub struct VrfOutputTruncated(pub Vec<u8>);
 
+impl_ref_hashable!(
+    VrfOutputTruncated,
+    append_bytes => 0
```
Comment (willemolding) [inline base/src/consensus_state.rs:25]: This macro is cute but I think it makes things less readable rather than more. It took me a while to figure out what is going on in this one for example.

Also it doesn't really save much space. I think this would be better without it

File: proof-systems-shim/src/lib.rs [hunk 6]
```diff
@@ -17,3 +17,43 @@
 pub use mina_hasher;
 pub use mina_signer;
 pub use o1_utils;
+
+/// macro that implements hashable trait for a reference value
+#[macro_export]
```
Comment (willemolding) [inline proof-systems-shim/src/lib.rs:22]: If we do have to include this macro I think it should live somewhere else. I think we should keep the shim crate as a pure shim (no additional code)

File: consensus/src/common.rs [hunk 7]
```diff
@@ -4,12 +4,16 @@
 //! Implements common APIs for the blockchain in the context of consensus.
```
Comment (willemolding) [inline consensus/src/common.rs:4]: This adds a bunch more of the difficult-to-read, repeated code tests we are trying to remove in #196. I don't think this is the way we should be writing these. Maybe chat with @creativcoder to see how they are planning to rewrite them and we can do it that way in the first place

**Review by willemolding — CHANGES_REQUESTED** [review]
Main point is to improve the strategy for writing tests in `consensus/src/common.rs`

File: base/src/consensus_state.rs (hunk 5 above)
Comment (sudo-shashank) [reply base/src/consensus_state.rs]: removed macro use for now

File: consensus/src/common.rs (hunk 7 above)
Comment (creativcoder) [reply consensus/src/common.rs:4]: Yes, we need not manually populate test payloads this way. For the same reason, I wanted to have our base internal types i.e., `ProtocolState` annotated with `Serialize`, `Deserialize`? @willemolding is it possible for us to add the Serialize annotation to those; would they break anything?

File: consensus/src/common.rs [hunk 8]
```diff
@@ -120,21 +124,23 @@ impl Chain<ProtocolState> for ProtocolStateChain {
...
-        self.top().map(|s| {
-            s.body
-                .consensus_state
-                .last_vrf_output
-                .hash()
-                .as_ref()
-                .encode_hex::<String>()
-        })
+    fn last_vrf_hash(&self) -> Option<Fp> {
+        static HASHER_POOL: OnceCell<SpinLockObjectPool<PoseidonHasherKimchi<VrfOutputTruncated>>> =
```
Comment (creativcoder) [inline consensus/src/common.rs:128]: Let's move this pool initialization block at the top as a function which returns `pool`. I see we initialize pool both in `state_hash()` and `last_vrf_hash()`

Comment (creativcoder) [inline consensus/src/common.rs:128]: Let's move this pool initialization block at the top as a function which returns `pool`. I see we initialize pool both in `state_hash()` and `last_vrf_hash()`

Comment (sudo-shashank) [reply consensus/src/common.rs:128]: hasher type is different for both, `state_hash()` uses ProtocolState and `last_vrf_hash()` uses VrfOutputTruncated

**Review by creativcoder — APPROVED** [review]

File: proof-systems-shim/src/lib.rs (hunk 6 above)
Comment (creativcoder) [reply proof-systems-shim/src/lib.rs]: Removed the macro as was not really helping much

File: consensus/src/common.rs (hunk 7 above)
Comment (creativcoder) [reply consensus/src/common.rs]: @willemolding  Update: We decided to remove this test and add it later as part of https://github.com/ChainSafe/mina-rs/issues/196 Additionally for removing the tech debt we'll need to have json serialization support from base types, which can be added after https://github.com/ChainSafe/mina-rs/pull/209 is merged. 
So merging this now.
